File: 238-product-of-array-except-self/solution.py
from typing import List
from functools import reduce


class Solution:
  def productExceptSelf(self, nums: List[int]) -> List[int]:
    num_zeroes, total_product = 0, 1
    for num in nums:
      if num == 0:
        num_zeroes += 1
      else:
        total_product *= num

    return [0 if num_zeroes >= 2 or (num_zeroes > 0 and num != 0)
            else total_product if num == 0
            else int(total_product / num)
            for num in nums]

File: 238-product-of-array-except-self/solution.cpp
#include <vector>

class Solution
{
public:
  std::vector<int> productExceptSelf(std::vector<int> &nums)
  {
    size_t num_zeroes = 0;
    int total_product = 1;
    for (const int num : nums)
    {
      if (num == 0)
      {
        ++num_zeroes;
      }
      else
      {
        total_product *= num;
      }
    }

    std::vector<int> result(nums.size(), 0);
    for (size_t i = 0; i < result.size(); ++i)
    {
      if (num_zeroes >= 2 || (num_zeroes > 0 && nums[i] != 0))
      {
        result[i] = 0;
      }
      else if (nums[i] == 0)
      {
        result[i] = total_product;
      }
      else
      {
        result[i] = total_product / nums[i];
      }
    }
    return result;
  }
};